Compiler failed while returning array for const init

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.5 22744ce.1e8d1a4
OS: macos, macOS, 13.6.1, 22G313
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz

getwd: /Users/rajesh/dev/buvi
vexe: /Users/rajesh/dev/v/v
vexe mtime: 2024-03-27 13:02:57

vroot: OK, value: /Users/rajesh/dev/v
VMODULES: OK, value: /Users/rajesh/.vmodules
VTMP: OK, value: /tmp/v_502

Git version: git version 2.39.3 (Apple Git-145)
Git vroot status: weekly.2024.13-24-g1e8d1a4d-dirty
.git/config present: true

CC version: Apple clang version 15.0.0 (clang-1500.1.0.2.5)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg ./bugs/bugs.v`
```v

const some_arr = make_array()

fn make_array() [64]u8 {
	mut arr := [64]u8{}
	return arr
}

```

**What did you expect to see?**

it should compile without any C errors.

**What did you see instead?**
```
==================
/tmp/v_502/bugs.01HT01E1J6178EZFQB8P470WH3.tmp.c:13008:24: error: assigning to '_v_Array_fixed_u8_64' (aka 'struct _v_Array_fixed_u8_64') from incompatible type 'u8[64]' (aka 'unsigned char[64]')
        _const_main__some_arr = main__make_array().ret_arr;
                              ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1731,6 +1731,7 @@ fn (mut c Checker) const_decl(mut node ast.ConstDecl) {
 		prev_const_var := c.const_var
 		c.const_var = unsafe { field }
 		mut typ := c.check_expr_option_or_result_call(field.expr, c.expr(mut field.expr))
+		typ = c.cast_fixed_array_ret(typ, c.table.sym(typ))
 		if ct_value := c.eval_comptime_const_expr(field.expr, 0) {
 			field.comptime_expr_value = ct_value
 			if ct_value is u64 {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5926,10 +5926,11 @@ fn (mut g Gen) const_decl_init_later(mod string, name string, expr ast.Expr, typ
 		}
 		if unwrap_option {
 			init.writeln(g.expr_string_surround('\t${cname} = *(${styp}*)', expr, '.data;'))
-		} else if (expr is ast.ArrayInit && (expr as ast.ArrayInit).has_index)
-			|| (expr is ast.CallExpr
-			&& g.table.final_sym(g.unwrap_generic((expr as ast.CallExpr).return_type)).kind == .array_fixed) {
+		} else if expr is ast.ArrayInit && (expr as ast.ArrayInit).has_index {
 			init.writeln(g.expr_string_surround('\tmemcpy(&${cname}, &', expr, ', sizeof(${styp}));'))
+		} else if expr is ast.CallExpr
+			&& g.table.final_sym(g.unwrap_generic((expr as ast.CallExpr).return_type)).kind == .array_fixed {
+			init.writeln(g.expr_string_surround('\tmemcpy(&${cname}, ', expr, ', sizeof(${styp}));'))
 		} else {
 			init.writeln(g.expr_string_surround('\t${cname} = ', expr, ';'))
 		}

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5926,7 +5926,9 @@ fn (mut g Gen) const_decl_init_later(mod string, name string, expr ast.Expr, typ
 		}
 		if unwrap_option {
 			init.writeln(g.expr_string_surround('\t${cname} = *(${styp}*)', expr, '.data;'))
-		} else if expr is ast.ArrayInit && (expr as ast.ArrayInit).has_index {
+		} else if (expr is ast.ArrayInit && (expr as ast.ArrayInit).has_index)
+			|| (expr is ast.CallExpr
+			&& g.table.final_sym(g.unwrap_generic((expr as ast.CallExpr).return_type)).kind == .array_fixed) {
 			init.writeln(g.expr_string_surround('\tmemcpy(&${cname}, &', expr, ', sizeof(${styp}));'))
 		} else {
 			init.writeln(g.expr_string_surround('\t${cname} = ', expr, ';'))

--- a/vlib/v/tests/const_array_test.v
+++ b/vlib/v/tests/const_array_test.v
@@ -11,7 +11,7 @@ fn make_array2() []u8 {
 	return arr
 }
 
-fn main() {
+fn test_main() {
 	assert some_arr.len == 64
 	assert some_arr2.len == 0
 }

--- a/vlib/v/tests/const_array_test.v
+++ b/vlib/v/tests/const_array_test.v
@@ -1,0 +1,17 @@
+const some_arr = make_array()
+const some_arr2 = make_array2()
+
+fn make_array() [64]u8 {
+	mut arr := [64]u8{}
+	return arr
+}
+
+fn make_array2() []u8 {
+	mut arr := []u8{}
+	return arr
+}
+
+fn main() {
+	assert some_arr.len == 64
+	assert some_arr2.len == 0
+}

